[ELF] Add BPSectionOrderer options

Colibrow opened the pull request:
Add new ELF linker options for profile-guided section ordering optimizations:

- `--irpgo-profile=<file>`: Read IRPGO profile data for use with startup and compression optimizations
- `--bp-startup-sort={none,function}`: Order sections based on profile data to improve star tup time
- `--bp-compression-sort={none,function,data,both}`: Order sections using balanced partitioning to improve compressed size
- `--bp-compression-sort-startup-functions`: Additionally optimize startup functions for compression
- `--verbose-bp-section-orderer`: Print statistics about balanced partitioning section ordering

Thanks to the @ellishg, @thevinster, and their team's work.

Colibrow commented:
@MaskRay @ellishg 
Hi, here is the ELF patch that adds support for the bp-* options. Could you please review it?

I also tested the ld64.lld build on my Mac locally using the --bp-compression-sort=function option to compare the effect of relocation hashes. Below are the results:
| file | size | gzipped size |
|------------|---------|---------|
| lld | 411,461,872 | 66,160,343 |
| lld-no-reloc | 411,426,304 | 65,935,315 |

In the lld-no-reloc version, I removed the following code responsible for calculating relocation hashes:
https://github.com/llvm/llvm-project/blob/1cc926b8b6976ac4a5a411eae564cfde2df1ef9d/lld/MachO/BPSectionOrderer.h#L104-L115
    
The results show that removing relocation hashes reduces the file size and gzip-compressed size slightly. I’m concerned that something might be wrong here—do you have any insights or thoughts on this?

MaskRay commented on lld/test/ELF/incompatible.s:
```diff
@@ -1,3 +1,16 @@
+# RUN: not ld.lld -o /dev/null --bp-startup-sort=function 2>&1 | FileCheck %s --check-prefix=STARTUP
+# STARTUP: --bp-startup-sort=function must be used with --irpgo-profile
+# RUN: not ld.lld -o /dev/null --bp-compression-sort-startup-functions 2>&1 | FileCheck %s --check-prefix=STARTUP-COMPRESSION
+# STARTUP-COMPRESSION: --bp-compression-sort-startup-functions must be used with --irpgo-profile
```
The test is for the specific error ` is incompatible with ...`

The negative tests should be moved to `lld/test/ELF/bp-section-orderer.s`, after the positive tests

MaskRay commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,134 @@
+      hashes.push_back(xxHash64(window));
+    }
+
+    // TODO: Calculate relocation hashes.
```
Delete TODO if it is likely we are not going to add relocations.

MaskRay commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,134 @@
+    size_t size = isec->content().size();
+    for (size_t i = 0; i < size; i++) {
+      auto window = isec->content().drop_front(i).take_front(windowSize);
+      hashes.push_back(xxHash64(window));
```
I am surprised that `xxHash64` doesn't need to be qualified.

This complex function should be moved toto BPSectionOrderer.cpp

why take_front windowSize?

Colibrow replied:
I moved `getSectionHashes` into BPSectionOrderer.cpp. I think take_front windowSize can construct the base hash unit for sliding window algorithm which lets Balanced Partitioning algorithm to infer if two functions are closer.

MaskRay commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,134 @@
+/// It is important that -ffunction-sections and -fdata-sections are used to
+/// ensure functions and data are in their own sections and thus can be
+/// reordered.
+llvm::DenseMap<const lld::elf::InputSectionBase *, int>
```
drop lld::elf:: inside namespace lld::elf

MaskRay commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,134 @@
+  llvm::StringRef getName() const override { return sym->getName(); }
+
+  const Defined *asDefined() const {
+    return llvm::dyn_cast_or_null<Defined>(sym);
```
drop _or_null if sym cannot be null

ellishg commented:
> @MaskRay @ellishg Hi, here is the ELF patch that adds support for the bp-* options. Could you please review it?
> 
> I also tested the ld64.lld build on my Mac locally using the --bp-compression-sort=function option to compare the effect of relocation hashes. Below are the results:
> 
> file	size	gzipped size
> lld	411,461,872	66,160,343
> lld-no-reloc	411,426,304	65,935,315
> In the lld-no-reloc version, I removed the following code responsible for calculating relocation hashes:
> 
> https://github.com/llvm/llvm-project/blob/1cc926b8b6976ac4a5a411eae564cfde2df1ef9d/lld/MachO/BPSectionOrderer.h#L104-L115
> 
> The results show that removing relocation hashes reduces the file size and gzip-compressed size slightly. I’m concerned that something might be wrong here—do you have any insights or thoughts on this?

Hmm I'm not sure why hashing relocations is not as effective on ELF. We know this is effective on our use cases on Mach-O and I think @dreampiggy measured some wins on Mach-O in [this comment](https://github.com/llvm/llvm-project/pull/96268#discussion_r1858053118). It might depend on some features of the binary you are building.

dreampiggy commented:
> I removed the following code responsible for calculating relocation hashes

This relocation hashes seems has a little benefit in both firefox-iOS Demo and our internal project. But I agreed that this may diff in different binaries. If possible, make this hashing calculation under an optional flag?

Colibrow commented:
@ellishg Yes. I believe it may depend on the binary features. After discussing with @dreampiggy, I've concluded that the relocation aspects likely have minimal impact. Therefore, I prefer to maintain the current implementation to avoid introducing additional complexity in ELF relocations. How do you think?

ellishg commented:
> @ellishg Yes. I believe it may depend on the binary features. After discussing with @dreampiggy, I've concluded that the relocation aspects likely have minimal impact. Therefore, I prefer to maintain the current implementation to avoid introducing additional complexity in ELF relocations. How do you think?

Seems ok for now. If someone finds that hashing relocations on ELF is useful in the future, they can look at the Mach-O implementation for reference and add it pretty easily

Colibrow commented:
@MaskRay Hi, Could you please review this PR? And Merry Christmas!

MaskRay commented on lld/test/ELF/bp-section-orderer.s:
```diff
@@ -0,0 +1,325 @@
+
+# STARTUP: Ordered 3 sections using balanced partitioning
+
+# RUN: ld.lld -e main -o - a.o --symbol-ordering-file a.orderfile --irpgo-profile=a.profdata --bp-startup-sort=function | llvm-nm --numeric-sort --format=just-symbols - | FileCheck %s --check-prefix=ORDERFILE
```
`_main` is Mach-O's entry. For ELF just use `_start` as the entry and remove `-e main`

MaskRay commented on lld/test/ELF/bp-section-orderer.s:
```diff
@@ -0,0 +1,325 @@
+# COMPRESSION-DATA: Ordered 7 sections using balanced partitioning
+# COMPRESSION-BOTH: Ordered 14 sections using balanced partitioning
+
+# RUN: not ld.lld -o /dev/null --bp-compression-sort=malformed 2>&1 | FileCheck %s --check-prefix=COMPRESSION-MALFORM
```
Since we have `cd %t` , `-o /dev/null` can be dropped

MaskRay commented on lld/test/ELF/bp-section-orderer.s:
```diff
@@ -0,0 +1,325 @@
+# RUN: not ld.lld -o /dev/null --bp-startup-sort=function 2>&1 | FileCheck %s --check-prefix=IRPGO-STARTUP
+# IRPGO-STARTUP: --bp-startup-sort=function must be used with --irpgo-profile
+# RUN: not ld.lld -o /dev/null --bp-compression-sort-startup-functions 2>&1 | FileCheck %s --check-prefix=STARTUP-COMPRESSION
+# STARTUP-COMPRESSION: --bp-compression-sort-startup-functions must be used with --irpgo-profile
```
Error message checks must start with `error: `

Colibrow commented:
@Andarwinux The IRPGO profile here is generated according to the [RFC](https://discourse.llvm.org/t/rfc-temporal-profiling-extension-for-irpgo/68068) and is designed for coherence with the Mach-O implementation. As of now, I believe it is unrelated to the lto-cs-profile-path and not well-suited for reuse. What do you think? cc @ellishg

ellishg commented:
I could be wrong but I think `lto-cs-profile-path` is used for profile generation with CSIRPGO, not profile consumption. Also, I think it isn't available for MachO so there would be a discrepancy there.

MaskRay commented on lld/test/ELF/bp-section-orderer-stress.s:
```diff
@@ -0,0 +1,104 @@
+# RUN: llvm-mc -filetype=obj -triple=aarch64 %t.s -o %t.o
+# RUN: llvm-profdata merge %t.proftext -o %t.profdata
+
+# RUN: ld.lld -e _start --icf=all -o - %t.o --irpgo-profile=%t.profdata --bp-startup-sort=function --bp-compression-sort-startup-functions --bp-compression-sort=both | llvm-nm --numeric-sort --format=just-symbols - > %t.order1.txt
```
`_start` is the default entry. Drop `-e _start` everywhere

MaskRay commented on lld/test/ELF/bp-section-orderer-stress.s:
```diff
@@ -0,0 +1,104 @@
+# REQUIRES: aarch64
+
+# Generate a large test case and check that the output is deterministic.
```
`## ` for non-RUN non-CHECK comments. Fix the other test as well

MaskRay commented on lld/test/ELF/incompatible.s:
```diff
@@ -1,3 +1,10 @@
+# REQUIRES: aarch64
+# RUN: rm -rf %t
+# RUN: not ld.lld -o /dev/null %t --irpgo-profile=/dev/null --bp-startup-sort=function --call-graph-ordering-file=/dev/null 2>&1 | FileCheck %s --check-prefix=IRPGO-ERR
```
This test is for incompatible relocatable files. Add such tests to bp-section-orderer.s

Please spend more time reading the existing tests to check whether the new tests fit there

MaskRay commented on lld/ELF/BPSectionOrderer.cpp:
```diff
@@ -0,0 +1,77 @@
+                                  bool forDataCompression,
+                                  bool compressionSortStartupFunctions,
+                                  bool verbose) {
+  size_t highestAvailablePriority = std::numeric_limits<int>::max();
```
Sent #121727 to remove `highestAvailablePriority`

Colibrow replied:
LGTM. I also feel that the implementation of priority in common seems a bit strange at first. I'll update this PR when https://github.com/llvm/llvm-project/pull/121727 merged.

MaskRay commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,115 @@
+    return isec->flags & llvm::ELF::SHF_EXECINSTR;
+  }
+
+  bool hasValidData() const override {
```
`hasValidData` is redundant. I've deleted it.

MaskRay commented:
@ellishg Hopefully I understand the features correctly...

### `--bp-compression-sort=` and `--bp-startup-sort=`

Both options instruct the linker to optimize section layout with the following goals:

* `--bp-compression-sort=[data|function|both]`: Improve Lempel-Ziv compression by grouping similar sections together, resulting in a smaller compressed app size.
* `--bp-startup-sort=function --irpgo-profile=<file>`: Utilize a temporal profile file to reduce page faults during program startup.

The linker determines the section order by considering three groups:

* Function sections ordered according to the temporal profile (`--irpgo-profile=`), prioritizing early-accessed and frequently accessed functions.
* Function sections. Sections containing similar functions are placed together, maximizing compression opportunities.
* Data sections. Similar data sections are placed together.

Within each group, the sections are ordered using the Balanced Partitioning algorithm.

The linker constructs a bipartite graph with two sets of vertices: sections and utility vertices.

* For profile-guided function sections:
  + The number of utility vertices is determined by the symbol order within the profile file.
  + If `--bp-compression-sort-startup-functions` is specified, extra utility vertices are allocated to prioritize nearby function similarity.
* For sections ordered for compression: Utility vertices are determined by analyzing k-mers of the section content and relocations.

The call graph profile is disables during this optimization.

When `--symbol-ordering-file=` is specified, sections described in that file are placed earlier.

MaskRay commented:
> If passing -flto -fprofile-use to clang, it also passes -plugin-opt=cs-profile-path to the linker. Therefore, it might be useful to standardize to irpgo-profile or reuse cs-profile-path to avoid duplicate options and simplify the use of this feature. (i.e. end-users only need to use -fprofile-use and -Wl,-bp-*)

> I could be wrong but I think lto-cs-profile-path is used for profile generation with CSIRPGO, not profile consumption. Also, I think it isn't available for MachO so there would be a discrepancy there. I can take a closer look Monday.

Second this. With the typical CSPGO usage lld --lto-cs-profile-path is for context-sensitive profile generation (`if (PGOOpt->CSAction == PGOOptions::CSIRInstr)`). 

```
# -fprofile-generate instruments the module
clang -O2 -c -flto=thin -fprofile-generate=profile a.c b.c
# -fprofile-generate just links in the runtime
clang -flto=thin -fprofile-generate=profile -fuse-ld=lld a.o b.o
# collect profile
./a.out
llvm-profdata merge profile/*.profraw -o default.profdata

# -fprofile-use performs optimization. -fcs-profile-generate at ThinLTO prelink phase just runs PGOInstrumentationGenCreateVar to create __llvm_profile_raw_version.
clang -O2 -c -flto=thin -fprofile-use=default.profdata -fcs-profile-generate=csprofile a.c b.c
# Pass -plugin-opt=cs-profile-path=csprofile/default_%m.profraw to lld.
# ThinLTO backend compile performs a post inline PGO instrumentation pass.
# Moreover, link in the runtime.
clang -flto=thin -fcs-profile-generate=csprofile -fuse-ld=lld a.o b.o
# Dump CS profile csprofile/default_%m.profraw
./a.out
# Merge it into the regular PGO profile
llvm-profdata merge default.profdata csprofile/*.profraw -o default.profdata

clang -O2 -c -flto=thin -fprofile-use=default.profdata a.c b.c
clang -flto=thin -fuse-ld=lld a.o b.o
# final executable
./a.out
```

It could also be utilized for CS profile use (`else if (PGOOpt->CSAction == PGOOptions::CSIRUse)`), but there isn't a proper lld test. @xur-llvm

Colibrow commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,115 @@
+
+  SmallVector<std::unique_ptr<BPSymbol>> getSymbols() const override {
+    SmallVector<std::unique_ptr<BPSymbol>> symbols;
+    if (auto *d = symbol->asDefined())
```
is there way to get all defined symbols from `InputSectionBase`? In MachO `isec->symbols` returns the symbols but I didn't find relative api in ELF.

Colibrow commented on lld/ELF/BPSectionOrderer.cpp:
```diff
@@ -0,0 +1,77 @@
+  // '--symbol-ordering-file' does.
+  SmallVector<std::unique_ptr<BPSectionBase>> sections;
+
+  for (Symbol *sym : ctx.symtab->getSymbols())
```
Could this be solved by using a unique `InputSectionBase*` set and retrieving symbols through `isec->file->getSymbols()` instead of matching symbols with sections one by one?

MaskRay replied:
Please don't resolve comments. The "resolve" button is for reviewers and trivial changes, not applicable here.

ellishg commented:
@Andarwinux It doesn't make sense to pass `--irpgo-profile=foo.profdata` (or use `--bp-compression-sort=both` `--bp-startup-sort=function`) during CSIRPGO generation (`-fcs-profile-generate`). These flags will only impact the function and data order in the final binary, which will not impact the resulting generated profiles. It only makes sense to use these to build your final optimized binary.

If we wanted to improve usability, we could have clang automatically pass `--irpgo-profile=foo.profdata` to the linker, but I don't think it makes sense to use the existing CSIRPGO linker flags.

Colibrow commented:
@MaskRay Hey! Made the changes you suggested:
- used DenseSet to avoid duplicate inputSections
- cleaned up test suite formatting
- updated getSymbols to return all defined symbols

Mind taking another look? Thanks!

MaskRay commented:
> @MaskRay Hey! Made the changes you suggested:
> 
> * used DenseSet to avoid duplicate inputSections
> * cleaned up test suite formatting
> * updated getSymbols to return all defined symbols
> 
> Mind taking another look? Thanks!

I am waiting on the approval of #121729 as a prerequisite. Then the symbol/section representation should be refined.

Colibrow committed:
Add new ELF linker options for order section layout

Colibrow committed:
try to refine symbol/section

Colibrow committed:
change content hash as #121729

Colibrow commented:
Hash changes have been updated like https://github.com/llvm/llvm-project/pull/121729. Not sure if the symbol/section representation is correct.

Colibrow commented:
gentle ping @MaskRay for any suggestions.

Colibrow committed:
remove useless header

MaskRay commented on lld/test/ELF/bp-section-orderer.s:
```diff
@@ -0,0 +1,328 @@
+# RUN: rm -rf %t && split-file %s %t && cd %t
+
+## Check for incompatible cases
+# RUN: not ld.lld -o /dev/null %t --irpgo-profile=/dev/null --bp-startup-sort=function --call-graph-ordering-file=/dev/null 2>&1 | FileCheck %s --check-prefix=BP-STARTUP-CALLGRAPH-ERR
```
Omit `-o /dev/null`

I can push commits to improve the test.

Colibrow replied:
Appreciate the suggestion.Done.

MaskRay commented:
I refactored BPSectionOrderer with C++ CRTP #124482 . For Mach-O, rootSymbolToSectionIdxs is computed when iterating sections. 

For ELF, it would be more efficient to iterate over symbols and populate rootSymbolToSectionIdxs. symbolToSectionIdxs can be removed.

Colibrow committed:
omit -o /dev/null

Colibrow committed:
rewrite BPSectionOrder in ELF following #124482

Colibrow commented:
Refactoring the ELF parts by following the https://github.com/llvm/llvm-project/pull/124482.

MaskRay commented on lld/ELF/BPSectionOrderer.cpp:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  std::optional<StringRef> static getResolvedLinkageName(llvm::StringRef name) {
+    return name;
```
`return {}`. We don't expect the call sites to use it.

Colibrow replied:
Done.

MaskRay commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,36 @@
+/// Run Balanced Partitioning to find the optimal function and data order to
+/// improve startup time and compressed size.
+///
+/// It is important that .subsections_via_symbols is used to ensure functions
```
Please don't blindly comments. ELF doesn't use subsections_via_symbols

Colibrow replied:
Sorry about that - I'll be more careful next time. Done.

MaskRay commented on lld/ELF/Options.td:
```diff
@@ -141,6 +141,24 @@ def call_graph_profile_sort: JJ<"call-graph-profile-sort=">,
 def : FF<"no-call-graph-profile-sort">, Alias<call_graph_profile_sort>, AliasArgs<["none"]>,
   Flags<[HelpHidden]>;
 
+defm irpgo_profile: Eq<"irpgo-profile",
```
EEq so that we don't define the single-dash `-irpgo-profile`

Colibrow replied:
Done.

MaskRay commented on lld/test/ELF/bp-section-orderer.s:
```diff
@@ -0,0 +1,328 @@
+const char*** r3 = &r2;
+const char* r4 = s2;
+void A() {
+    return;
```
2-space indentation

Colibrow replied:
re-format the file. Done.

Colibrow committed:
fix: address code review feedback

Colibrow commented on lld/test/ELF/bp-section-orderer-stress.s:
```diff
@@ -0,0 +1,104 @@
+# RUN: llvm-mc -filetype=obj -triple=aarch64 %t.s -o %t.o
+# RUN: llvm-profdata merge %t.proftext -o %t.profdata
+
+# RUN: ld.lld --icf=all -o - %t.o --irpgo-profile=%t.profdata --bp-startup-sort=function --bp-compression-sort-startup-functions --bp-compression-sort=both | llvm-nm --numeric-sort --format=just-symbols - > %t.order1.txt
```
Done.

MaskRay committed:
[cmake] Add ProfileData to fix BUILD_SHARED_LIBS=on builds

MaskRay committed:
Simplify

MaskRay committed:
Ignore callGraphProfile when BP orderer is used, update test

MaskRay committed:
Move getResolvedLinkageName to .inc

MaskRay committed:
Replace Symbol with Defined

MaskRay committed:
Optimize getSymbols

MaskRay commented on lld/ELF/BPSectionOrderer.cpp:
```diff
@@ -0,0 +1,95 @@
+  static bool isCodeSection(const Section &sec) {
+    return sec.flags & llvm::ELF::SHF_EXECINSTR;
+  }
+  ArrayRef<Defined *> getSymbols(const Section &sec) {
```
The previous getSymbols iterated over all symbols in sec.file, which could be very slow. I rewrote this to use a member variable.

MaskRay commented on lld/ELF/CMakeLists.txt:
```diff
@@ -72,6 +73,7 @@ add_lld_library(lldELF
   Object
   Option
   Passes
+  ProfileData
```
This is needed to fix BUILD_SHARED_LIBS=on builds

MaskRay committed:
improve test
